torch/distributed/nn/api/remote_module.py

```diff
@@ -0,0 +1,245 @@
+        >>> import torch
+        >>> import torch.distributed.rpc as rpc
+        >>>
+        >>> rpc.init_rpc("worker1", rank=1, world_size=2)
+        >>> rpc.shutdown()
+    """
+    # Sanity checks.
+    assert rpc._is_current_rpc_agent_set(), "RemoteModule only works in RPC."
+
+    # Default arguments preperation.
+    args = args if args is not None else ()
+    kwargs = kwargs if kwargs is not None else {}
+    global_unique_name = (  # Assign a global name for the module to be created.
+        global_unique_name
+        if global_unique_name is not None
+        else _gen_global_unique_name()
+    )
+
+    if module_interface_cls is not None:
+        is_scriptable = True
+
+        # Instantiate template on remote side.
+        fut = rpc.rpc_async(
+            to, _instantiate_template, (module_interface_cls,)
+        )
+
+        # Instantiate template on local side.
+        generated_module = instantiator.instantiate_scriptable_remote_module_template(module_interface_cls)
+        remote_module_cls = generated_module._RemoteModule
+
+        # Create the module on the remote side.
+        fut.wait()  # Ensure remote_module_cls is available on remote side.
+        module_rref = rpc.rpc_sync(
+            to,
+            _scripted_module_creator_wrapper,
+            (module_creator, module_interface_cls, args, kwargs),
+        )
+    else:
+        is_scriptable = False
+        remote_module_cls = _NON_SCRIPTABLE_REMOTE_MODULE_CLS
+
+        # Create the module on the remote side.
+        module_rref = rpc.remote(to, module_creator, args, kwargs)
+
+    # Create remote_module_cls instance on local side.
+    remote_module = remote_module_cls(module_rref, is_scriptable, global_unique_name)
+
+    return remote_module
+
+
+def RemoteModule(
+    to: str,
+    module_creator: Callable,
+    args: Tuple = None,
+    kwargs: Dict[str, Any] = None,
+    global_unique_name: str = None,
+):
+    """
+        A RemoteModule instance can only be created after RPC initialization.
+        It creates a user-specified module on a specified remote node.
+        It behaves like a regular ``nn.Module`` except that the ``forward`` method is
+        executed on the remote node.
+        It takes care of autograd recording to ensure the backward pass propogates
+        gradients back to the corresponding remote module.
+
+        The arguments of ``forward_async`` and ``forward`` are the same as
+        the ``forward`` method of the module returned by the ``module_creator``.
+
+        For example, if ``module_creator`` returns an instace of ``nn.Linear``,
+        that has ``forward`` method signature, ``def forward(input: Tensor) -> Tensor:``,
+        the generated ``RemoteModule`` will have 2 methods in signature of
+        ``def forward(input: Tensor) -> Tensor:`` and
+        ``def forward_async(input: Tensor) -> Future[Tensor]:``.
+
+    Arguments:
+        to (str or WorkerInfo): id or name of the destination worker.
```

Comment:
Nit: just noticed this, maybe we should call this `on` instead and mention in the docs `id or name of the worker on which this module should be created`. `on` is probably more accurate here since we're creating something **on** a worker instead of sending something **to** a worker when using rpc_async etc.

Reply:
Sounds better. Updating.